File: samples/cpp/tutorial_code/HighGUI/AddingImagesTrackbar.cpp
/**
 * @file AddingImagesTrackbar.cpp
 * @brief Simple linear blender ( dst = alpha*src1 + beta*src2 )
 * @author OpenCV team
 */

#include "opencv2/imgcodecs.hpp"
#include "opencv2/highgui.hpp"
#include <iostream>

using namespace cv;
using std::cout;

/** Global Variables */
const int alpha_slider_max = 100;
int alpha_slider;
double alpha;
double beta;

/** Matrices to store images */
Mat src1;
Mat src2;
Mat dst;

//![on_trackbar]
/**
 * @function on_trackbar
 * @brief Callback for trackbar
 */
static void on_trackbar(int pos, void* userdata) {
    // Print trackbar position
    std::cout << "Trackbar position: " << pos << std::endl;
    // If userdata is provided, use it
    if (userdata) {
        int* user_data = static_cast<int*>(userdata);
        std::cout << "User data: " << *user_data << std::endl;
    }
    // Calculate alpha and beta values
    alpha = (double)pos / alpha_slider_max;
    beta = (1.0 - alpha);
    addWeighted(src1, alpha, src2, beta, 0.0, dst);
    imshow("Linear Blend", dst);
}
//![on_trackbar]

/**
 * @function main
 * @brief Main function
 */
int main(void)
{
    //![load]
    /// Read images (both must be of the same size and type)
    src1 = imread(samples::findFile("LinuxLogo.jpg"));
    src2 = imread(samples::findFile("WindowsLogo.jpg"));
    //![load]

    if (src1.empty()) { cout << "Error loading src1 \n"; return -1; }
    if (src2.empty()) { cout << "Error loading src2 \n"; return -1; }

    // Initialize trackbar value
    alpha_slider = 0;

    //![window]
    namedWindow("Linear Blend", WINDOW_AUTOSIZE); //Create Window
    //![window]

    //![create_trackbar]
    char TrackbarName[50];
    snprintf(TrackbarName, sizeof(TrackbarName), "Alpha x %d", alpha_slider_max);
    // Example userdata: Pass a pointer to an integer as userdata
    int user_value = 42; // Arbitrary value for demonstration
    createTrackbar(TrackbarName, "Linear Blend", &alpha_slider, alpha_slider_max, on_trackbar, &user_value);
    //![create_trackbar]

    /// Show initial result
    on_trackbar(alpha_slider, nullptr);

    /// Wait for user input
    waitKey(0);
    return 0;
}

File: samples/python/tutorial_code/highgui/trackbar/AddingImagesTrackbar.py
from __future__ import print_function
from __future__ import division
import cv2 as cv
import argparse

alpha_slider_max = 100
title_window = 'Linear Blend'

## [on_trackbar]
def on_trackbar(val):
    alpha = val / alpha_slider_max
    beta = ( 1.0 - alpha )
    dst = cv.addWeighted(src1, alpha, src2, beta, 0.0)
    cv.imshow(title_window, dst)
## [on_trackbar]

parser = argparse.ArgumentParser(description='Code for Adding a Trackbar to our applications tutorial.')
parser.add_argument('--input1', help='Path to the first input image.', default='LinuxLogo.jpg')
parser.add_argument('--input2', help='Path to the second input image.', default='WindowsLogo.jpg')
args = parser.parse_args()

## [load]
# Read images ( both have to be of the same size and type )
src1 = cv.imread(cv.samples.findFile(args.input1))
src2 = cv.imread(cv.samples.findFile(args.input2))
## [load]
if src1 is None:
    print('Could not open or find the image: ', args.input1)
    exit(0)

if src2 is None:
    print('Could not open or find the image: ', args.input2)
    exit(0)

## [window]
cv.namedWindow(title_window)
## [window]

## [create_trackbar]
trackbar_name = 'Alpha x %d' % alpha_slider_max
cv.createTrackbar(trackbar_name, title_window , 0, alpha_slider_max, on_trackbar)
## [create_trackbar]

# Show some stuff
on_trackbar(0)

# Wait until user press some key
cv.waitKey()
